### DISCUSSION TITLE
Setting datepicker's ViewDate from Bean

### DISCUSSION THREAD

**Discussion by jncolin** [primefaces/community#2196] [PrimeFaces] [answered]
Hi
I'm struggling with the datepicker calendar, using PF14.0.1. 

My scenario is simple: I have a form with a datepicker, I can select one or multiple dates, click 'Add' and this fires an action in the bean. At the end of the action, a growl message is displayed. 

I am using a multi-select datepicker component, with _value_ pointing to a List in the backing bean. This list contains the selected dates, so when I need to reset the datepicker (i.e. when I return from the action), the list is null. The issue is that in this case, when the component is refreshed, it always displays the current month.

Why isn't there an attribute in the component that allows to set the viewdate? I know it can be done in JavaScript, but then, it interferes with the growl message. 

Thanks for your help

Jean

**Comment by vieira-brz** [comment]
Hi Jean,

It looks like PrimeFaces doesn’t have a built-in `viewDate` attribute for the `<p:calendar>` component that you can use directly from the server-side, but you can tackle this with a bit of JavaScript and AJAX. Here’s how to do it without messing with your growl notifications:

1. **JavaScript Function:**
   First, create a simple JavaScript function that sets the calendar's view date:
   ```javascript
   function setCalendarViewDate(date) {
       PF('yourWidgetVar').setViewDate(date);
   }
   ```

2. **AJAX Call:**
   Add an `oncomplete` attribute in your 'Add' button to call this JavaScript function after the server-side action is complete. You can specify any date you want the calendar to reset to:
   ```xhtml
   <p:commandButton value="Add" actionListener="#{yourBean.yourAction}"
                    oncomplete="setCalendarViewDate(new Date(2023, 0, 1))" update="calendarId growlId"/>
   ```

   Change `new Date(2023, 0, 1)` to whichever year, month (remember, January is 0!), and day you need.

3. **Setup Your Calendar:**
   Make sure your calendar allows multiple selections and is connected to the right bean property:
   ```xhtml
   <p:calendar widgetVar="yourWidgetVar" value="#{yourBean.dateList}" selectionMode="multiple" id="calendarId"/>
   ```

4. **Growl Notification:**
   Update your growl as usual:
   ```xhtml
   <p:growl id="growlId" life="3000"/>
   ```

If you find that the JavaScript messes with your growl messages, check the timing of the script execution in `oncomplete`. You might need to tweak how they sequence or even delay the JavaScript call slightly with a timeout if needed.

This way, you can manually reset your calendar to a specific view date straight from your action method, using both the client-side capabilities of PrimeFaces and your server control.

**Reply by melloware** [reply]
See this ticket to fix it: https://github.com/primefaces/primefaces/issues/12157

**Reply by jncolin** [reply]
Hi @vieira-brz and @melloware 
Actually, I am following the approach proposed by @vieira-brz ; however, it seems that changing the date in the JS also triggers the viewChanged event on the datepicker, and since the dateList is empty, the date is then reset to the current date.
Thanks for your help

**Reply by jncolin** [reply]
@vieira-brz I have prepared a small test app; when you run it, go to http://localhost:5000/playground.xhtml
there's a datepicker, a message and a button. The datepicker has two listeners for the viewChange and dateSelect events, and both update the message and the growl area. Both work fine. The trouble comes with the button, that fires a bean action. 
The bean has two fields, one to record the list of selecteddates (initially empty, and cleared when button action fires), and a currentDate, initially set to now() and updated in the viewchange listener. 
When you click the button, the list of selectedDates is cleared; when action returns, I would expect it to stay on the currentMonth, since I'm calling JS script code for that purpose in the oncomplete handler. 
The issue is that right after the oncomplete function is called, a new viewChange event is triggered (no idea why, except because of the JS call), and this resets the datepicker to today's date.

To reproduce, scroll to november, select one or two dates, (you see messages along your actions), then click the button, the growl message does not show (actually it does but it is immediately overwritten by that from the viewChange listener) and the datepicker goes back to June.

All ideas welcome :-) Thanks a lot
[primefaces-test-jakarta.tar.gz](https://github.com/user-attachments/files/15968515/primefaces-test-jakarta.tar.gz)

**Reply by melloware** [reply]
@jncolin yes if you use `ajax="ViewChange"` you will get an infinite loop if you call the JavaScript method `PF('yourWidgetVar').setViewDate(date);` as that method triggers the AJAX view date change.  So the only way you can use `PF('yourWidgetVar').setViewDate(date);` is if you DON'T use p:ajax event="viewChange"

**Reply by jncolin** [reply]
@melloware thanks for the explanations. what is most important for me is that viewChange is properly handled, since other elements of the page must be updated every time the view changes; and ideally, after an action, I should come back to the same view, not to today. Any hint? 
Thanks a lot

**Reply by melloware** [reply, marked as answer]
I fixed your code:
[discussion-2196.zip](https://github.com/user-attachments/files/15972148/discussion-2196.zip)

2 things.

1. you were not updating the Script element so it was only loaded on page load as current date so you needed to update it to be the correct date after the AJAX request.
2. I hacked in test.js the DatePicker not to fire the `viewChange` AJAX if you are calling it manually.

**Reply by jncolin** [reply]
> I fixed your code: [discussion-2196.zip](https://github.com/user-attachments/files/15972148/discussion-2196.zip)
> 
> 2 things.
> 
> 1. you were not updating the Script element so it was only loaded on page load as current date so you needed to update it to be the correct date after the AJAX request.
> 2. I hacked in test.js the DatePicker not to fire the `viewChange` AJAX if you are calling it manually.

Wonderful, this is working, I can freely move the datepicker, perform action and stay on the same month view, even getting the growl messages. I think my issue is solved, thanks so much

**Reply by melloware** [reply]
I think adding `setViewDate(date, silent)` would be a good PF addition.  I will create a ticket.

**Reply by melloware** [reply]
Done: https://github.com/primefaces/primefaces/issues/12185
